src/Features/Core/Portable/CodeRefactorings/AbstractRefactoringHelpersService.cs

```diff
@@ -0,0 +1,176 @@
+
+            // Handle selections
+            // - the smallest node whose span inlcudes whole selection
+            // - extraction from such node is attempted, result tested via predicate
+            // - travels upwards through same-sized nodes, extracting and testing predicate
+            var node = root.FindNode(selectionStripped, getInnermostNodeForTie: true);
+            SyntaxNode prevNode;
+            do
+            {
+                var wantedNode = TryGetAcceptedNodeOrExtracted(node, predicate, extractNode, syntaxFacts);
+                if (wantedNode != default)
+                {
+                    return wantedNode;
+                }
+
+                prevNode = node;
+                node = node.Parent;
+            }
+            while (node != null && prevNode.FullWidth() == node.FullWidth());
+
+            // consider what selection is touching only when it's empty -> prevents
+            // `[|C|] methodName(){}` from registering as relevant for method Node
+            if (!selection.IsEmpty)
+            {
+                return default;
+            }
+
+            // get Token for current selection (empty) location
+            var tokenOnSelection = root.FindToken(selectionStripped.Start);
+
+            // Token to right or containing current selection
+            var tokenToRightOrIn = tokenOnSelection.Span.Contains(selectionStripped.Start)
+                ? tokenOnSelection
+                : default;
+
+            if (tokenToRightOrIn != default)
+            {
+                var rightNode = tokenOnSelection.Parent;
+                do
+                {
+                    // consider either a Node that is a parent of touched Token (selection can be within) or ancestor Node of such Token whose span starts on selection
+                    var wantedNode = TryGetAcceptedNodeOrExtracted(rightNode, predicate, extractNode, syntaxFacts);
+                    if (wantedNode != default)
+                    {
+                        return wantedNode;
+                    }
+
+                    rightNode = rightNode?.Parent;
+                }
+                while (rightNode != null && rightNode.Span.Start == selection.Start);
+            }
+
+            // if selection inside tokenToRightOrIn -> no Token can be to Left (tokenToRightOrIn is left from selection)
+            if (tokenToRightOrIn != default && tokenToRightOrIn.Span.Start != selectionStripped.Start)
+            {
+                return default;
+            }
+
+            // Token to left
+            var tokenPreSelection = (tokenOnSelection.Span.End == selectionStripped.Start)
+                ? tokenOnSelection
+                : tokenOnSelection.GetPreviousToken();
+
+            var tokenToLeft = (tokenPreSelection.Span.End == selectionStripped.Start)
+                ? tokenPreSelection
+                : default;
+
+            if (tokenToLeft != default)
```

Comment:
i think it would be valuable to have a comment explaining what this entire block is trying to accomplish at a conceptual level.